mitxgraders/helpers/calc.py

```diff
@@ -549,23 +448,331 @@ def check_variables(self, valid_variables, valid_functions):
+                msg = ("There was an error evaluating {name}(...). "
+                       "Its input does not seem to be in its domain.").format(name=name)
+                raise FunctionEvalError(msg)
+
+    def eval_parens(self, parse_result):
+        """
+        A wrapper that returns the first element of parse_result. Used for `parentheses`.
+
+        Arguments:
+            parse_result: A list containing a single result containing the evaluation
+            of the expression in parentheses
+
+        Usage
+        =====
+        >>> parser = FormulaParser("1", {"%": 0.01})
+        >>> parser.eval_parens([1])
+        1
+        """
+        return parse_result[0]
+
+    def eval_atom(self, parse_result):
+        """
+        A wrapper that returns the value of the wrapped atom. Used for `atom`.
+
+        Arguments:
+            parse_result: A list containing a single result containing the evaluation
+            of the atom
+
+        Usage
+        =====
+        >>> parser = FormulaParser("1", {"%": 0.01})
+        >>> parser.eval_atom([1])
+        1
+        """
+        return parse_result[0]
+
+    def eval_power(self, parse_result):
+        """
+        Take a list of numbers and exponentiate them, right to left.
+
+        Can also have minus signs interspersed, which means to flip the sign of the
+        exponent.
+
+        Arguments:
+            parse_result: A list of numbers and "-" strings, to be read as exponentiated
+            [a, b, c, d] = a^b^c^d
+            [a, "-", c, d] = a^(-(c^d))
+
+        Usage
+        =====
+        >>> parser = FormulaParser("1", {"%": 0.01})
+        >>> parser.eval_power([4,3,2])  # Evaluate 4^3^2
+        262144
+        >>> parser.eval_power([2,"-",2,2])  # Evaluate 2^(-(2^2))
+        0.0625
+        """
+        def robust_pow(base, exponent):
+            """The power function that we want to use"""
+            # This will need to be modified if base can be a matrix
+            try:
+                return base ** exponent
+            except ValueError:
+                return np.lib.scimath.power(base, exponent)
+
+        data = parse_result[:]
+        result = data.pop()
+        while data:
+            # Result contains the current exponent
+            working = data.pop()
+            if working == "-":
+                result *= -1
+            else:
+                # working is base, result is exponent
+                result = robust_pow(working, result)
+
+        return result
+
+    def eval_negation(self, parse_result):
+        """
+        Negate a number an appropriate number of times.
+
+        Arguments:
+            parse_result: A list containing zero or more "-" strings, followed by a number.
+            ["-", "-", "-", a] = -a
+            ["-", "-", a] = a
+
+        Usage
+        =====
+        >>> parser = FormulaParser("1", {"%": 0.01})
+        >>> parser.eval_negation([2])
+        2
+        >>> parser.eval_power(["-",2])
```

Comment:
Should this (and the next few) be `eval_negation`? (Or, `eval_sign` if you take the suggestion above.)